File: packages/kits/LocusKit/Tests/LocusKitTests/SensitivityInheritanceTests.swift
// SensitivityInheritanceTests.swift
//
// Sensitivity inheritance (§D.1, §D.2) at the LocusKit layer:
//
//   - consolidateTransactionally stamps each _consolidated_from tunnel with
//     max(vague, constituent) adjective sensitivity (bits 6–11, cookbook §2.3).
//   - foldInTransactionally stamps the supersedes tunnel with v2's tier and
//     each _consolidated_from tunnel with max(v2, constituent) tier.
//   - propose stamps the proposal's own adjectiveBitmap bits 6–11 with the
//     target drawer's adjective sensitivity (§D.2).
//
// All tests operate at the DrawerStore / Estate layer; no GLK required.

import Foundation
import Testing
import SubstrateKernel
@testable import LocusKit

@Suite("SensitivityInheritanceTests — tunnel stamp + proposal stamp (§D.1, §D.2)")
struct SensitivityInheritanceTests {

    // MARK: - Fixture helpers

    private func t(_ epoch: TimeInterval) -> Date {
        Date(timeIntervalSince1970: epoch)
    }

    private func makeTempURL() -> URL {
        let name = "locuskit-sensitest-\(UUID().uuidString).sqlite"
        return URL(fileURLWithPath: NSTemporaryDirectory()).appendingPathComponent(name)
    }

    private func cleanup(_ url: URL) {
        try? FileManager.default.removeItem(at: url)
        try? FileManager.default.removeItem(at: url.appendingPathExtension("sqlite-wal"))
        try? FileManager.default.removeItem(at: url.appendingPathExtension("sqlite-shm"))
    }

    private func makeStore() async throws -> (DrawerStore, URL) {
        let url = makeTempURL()
        let store = try await DrawerStore(storage: TestStorage.sqlite(url))
        return (store, url)
    }

    /// Create a plain episodic drawer with a given adjective sensitivity.
    private func episodicDrawer(id: String, sensitivity: AdjectiveSensitivity) -> Drawer {
        let adjBitmap = BitField.writeField(
            Int64(sensitivity.rawValue), into: 0, shift: 6, width: 6)
        return Drawer(
            id: TestStorage.tid(id),
            content: "episodic content \(id)",
            parentNodeId: "test-parent",
            addedBy: "newton",
            filedAt: t(1_700_000_000),
            embeddingModelID: "test-model-v1",
            adjectiveBitmap: adjBitmap
        )
    }

    /// Create a vague drawer with isVague (bit 20) set and a given sensitivity.
    private func vagueDrawer(id: String, sensitivity: AdjectiveSensitivity) -> Drawer {
        let adjBitmap = BitField.writeField(
            Int64(sensitivity.rawValue), into: 0, shift: 6, width: 6)
        return Drawer(
            id: TestStorage.tid(id),
            content: "vague summary \(id)",
            parentNodeId: "test-parent",
            addedBy: "newton",
            filedAt: t(1_700_001_000),
            embeddingModelID: "test-model-v1",
            adjectiveBitmap: adjBitmap,
            operationalBitmap: DrawerFeatureFlags.isVague.rawValue
        )
    }

    // MARK: - consolidateTransactionally tunnel stamp

    @Test("consolidateTransactionally stamps _consolidated_from tunnels with max(vague, constituent) sensitivity")
    func consolidate_tunnelsCarryMaxSensitivity() async throws {
        let (store, url) = try await makeStore()
        defer { cleanup(url) }

        // One normal constituent and one restricted constituent;
        // vague drawer is at normal. Tunnel to the restricted constituent
        // must carry restricted; tunnel to the normal constituent inherits
        // the vague tier (normal).
        let d1 = episodicDrawer(id: "d1", sensitivity: .normal)
        let d2 = episodicDrawer(id: "d2", sensitivity: .restricted)
        let d3 = episodicDrawer(id: "d3", sensitivity: .normal)
        for d in [d1, d2, d3] { try await store.addDrawer(d) }

        let vague = vagueDrawer(id: "v1", sensitivity: .normal)
        try await store.consolidateTransactionally(
            vagueDrawer: vague,
            constituentIDs: [d1.id, d2.id, d3.id],
            addedBy: "newton",
            now: t(1_700_002_000)
        )

        // Fetch all tunnels for this consolidation.
        let allTunnels = try await store.tunnelsTouching(vague.id)

        // Tunnel to the restricted constituent must carry restricted sensitivity.
        let tunnelToD2 = allTunnels.first { $0.targetDrawerId == d2.id }
        let t2 = try #require(tunnelToD2, "tunnel to restricted constituent must exist")
        #expect(t2.adjectiveSensitivity == .restricted,
                "tunnel to restricted constituent must carry .restricted tier")

        // Tunnel to normal constituent inherits vague tier (normal).
        let tunnelToD1 = allTunnels.first { $0.targetDrawerId == d1.id }
        let t1 = try #require(tunnelToD1, "tunnel to normal constituent must exist")
        #expect(t1.adjectiveSensitivity == .normal,
                "tunnel to normal constituent inherits vague tier (.normal)")
    }

    @Test("consolidateTransactionally: vague drawer elevated lifts all tunnel tiers to elevated")
    func consolidate_vagueElevatedLiftsAllTunnels() async throws {
        let (store, url) = try await makeStore()
        defer { cleanup(url) }

        let d1 = episodicDrawer(id: "d1", sensitivity: .normal)
        let d2 = episodicDrawer(id: "d2", sensitivity: .normal)
        let d3 = episodicDrawer(id: "d3", sensitivity: .normal)
        for d in [d1, d2, d3] { try await store.addDrawer(d) }

        // Vague drawer stamped elevated (e.g. repair sweep scenario).
        let vague = vagueDrawer(id: "v1", sensitivity: .elevated)
        try await store.consolidateTransactionally(
            vagueDrawer: vague,
            constituentIDs: [d1.id, d2.id, d3.id],
            addedBy: "newton",
            now: t(1_700_002_000)
        )

        let allTunnels = try await store.tunnelsTouching(vague.id)
        // All three tunnels must carry elevated (from the vague drawer).
        let cfTunnels = allTunnels.filter { $0.label == "_consolidated_from" }
        #expect(cfTunnels.count == 3)
        for tun in cfTunnels {
            #expect(tun.adjectiveSensitivity == .elevated,
                    "vague drawer at .elevated must lift tunnel tier to .elevated")
        }
    }

    // MARK: - foldInTransactionally tunnel stamp

    @Test("foldInTransactionally supersedes tunnel carries v2 sensitivity")
    func foldIn_supersedesTunnelCarriesV2Sensitivity() async throws {
        let (store, url) = try await makeStore()
        defer { cleanup(url) }

        // Baseline: 3 normal constituents, vague v1 at normal.
        let d1 = episodicDrawer(id: "d1", sensitivity: .normal)
        let d2 = episodicDrawer(id: "d2", sensitivity: .normal)
        let d3 = episodicDrawer(id: "d3", sensitivity: .normal)
        for d in [d1, d2, d3] { try await store.addDrawer(d) }
        let v1 = vagueDrawer(id: "v1", sensitivity: .normal)
        try await store.consolidateTransactionally(
            vagueDrawer: v1,
            constituentIDs: [d1.id, d2.id, d3.id],
            addedBy: "newton",
            now: t(1_700_002_000)
        )

        // A fourth normal constituent folds in; v2 is stamped elevated
        // (e.g. it absorbed a new elevated constituent at the GLK layer).
        let d4 = episodicDrawer(id: "d4", sensitivity: .elevated)
        try await store.addDrawer(d4)
        let v2 = vagueDrawer(id: "v2", sensitivity: .elevated)
        // Share lineageID with v1 (required by foldIn precondition).
        let v2WithLineage = Drawer(
            id: TestStorage.tid("v2"),
            content: "vague v2",
            parentNodeId: "test-parent",
            addedBy: "newton",
            filedAt: t(1_700_003_000),
            embeddingModelID: "test-model-v1",
            adjectiveBitmap: BitField.writeField(
                Int64(AdjectiveSensitivity.elevated.rawValue), into: 0, shift: 6, width: 6),
            operationalBitmap: DrawerFeatureFlags.isVague.rawValue,
            lineageID: v1.lineageID
        )
        try await store.foldInTransactionally(
            vagueV2: v2WithLineage,
            priorVagueID: v1.id,
            enlargedConstituentIDs: [d1.id, d2.id, d3.id, d4.id],
            addedBy: "newton",
            now: t(1_700_004_000)
        )

        // Supersedes tunnel must carry v2's elevated tier.
        let allTunnels = try await store.tunnelsTouching(v2WithLineage.id)
        let supersedesTun = allTunnels.first { $0.kind == .supersedes }
        let st = try #require(supersedesTun, "supersedes tunnel must exist")
        #expect(st.adjectiveSensitivity == .elevated,
                "supersedes tunnel must carry v2's .elevated sensitivity")
    }

    @Test("foldInTransactionally _consolidated_from tunnels carry max(v2, constituent) sensitivity")
    func foldIn_cfTunnelsCarryMaxSensitivity() async throws {
        let (store, url) = try await makeStore()
        defer { cleanup(url) }

        let d1 = episodicDrawer(id: "d1", sensitivity: .normal)
        let d2 = episodicDrawer(id: "d2", sensitivity: .restricted)
        let d3 = episodicDrawer(id: "d3", sensitivity: .normal)
        for d in [d1, d2, d3] { try await store.addDrawer(d) }
        let v1 = vagueDrawer(id: "v1", sensitivity: .normal)
        try await store.consolidateTransactionally(
            vagueDrawer: v1,
            constituentIDs: [d1.id, d2.id, d3.id],
            addedBy: "newton",
            now: t(1_700_002_000)
        )

        // v2 is at normal; constituent d2 is restricted → d2's tunnel must be restricted.
        let d4 = episodicDrawer(id: "d4", sensitivity: .normal)
        try await store.addDrawer(d4)
        let v2WithLineage = Drawer(
            id: TestStorage.tid("v2"),
            content: "vague v2",
            parentNodeId: "test-parent",
            addedBy: "newton",
            filedAt: t(1_700_003_000),
            embeddingModelID: "test-model-v1",
            operationalBitmap: DrawerFeatureFlags.isVague.rawValue,
            lineageID: v1.lineageID
        )
        try await store.foldInTransactionally(
            vagueV2: v2WithLineage,
            priorVagueID: v1.id,
            enlargedConstituentIDs: [d1.id, d2.id, d3.id, d4.id],
            addedBy: "newton",
            now: t(1_700_004_000)
        )

        let allTunnels = try await store.tunnelsTouching(v2WithLineage.id)
        let tunnelToD2 = allTunnels.first {
            $0.label == "_consolidated_from" && $0.targetDrawerId == d2.id
        }
        let t2 = try #require(tunnelToD2, "tunnel to restricted constituent must exist")
        #expect(t2.adjectiveSensitivity == .restricted,
                "_consolidated_from tunnel to restricted constituent must carry .restricted")

        let tunnelToD1 = allTunnels.first {
            $0.label == "_consolidated_from" && $0.targetDrawerId == d1.id
        }
        let t1 = try #require(tunnelToD1)
        #expect(t1.adjectiveSensitivity == .normal,
                "_consolidated_from tunnel to normal constituent carries .normal when v2 is normal")
    }

    // MARK: - proposal sensitivity stamp (§D.2)

    @Test("propose stamps proposal adjectiveBitmap bits 6–11 with target drawer's sensitivity")
    func propose_inheritsTargetSensitivity() async throws {
        let dir = FileManager.default.temporaryDirectory
            .appendingPathComponent("locuskit-prop-sens-\(UUID().uuidString)")
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        let path = dir.appendingPathComponent("estate.sqlite3")
        defer { try? FileManager.default.removeItem(at: dir) }

        // Capture a restricted drawer and propose against it.
        let estate = try await Estate.create(
            storage: TestStorage.sqlite(path),
            owner: OwnerCredentials(ownerIdentifier: "test-owner")
        )
        let frame = CaptureFrame(
            content: "restricted knowledge",
            channel: .typed,
            room: "test-room",
            latticeAnchor: LatticeAnchor(udcCode: "005"),
            addedBy: "test-agent",
            embeddingModelID: "minilm-v1",
            sensitivity: .restricted
        )
        let target = try await estate.capture(frame)
        #expect(target.adjectiveSensitivity == .restricted,
                "target drawer must carry .restricted tier")

        let proposal = try await estate.propose(
            ProposeFrame(
                target: target.id,
                kind: .mutateDrawer,
                justification: "test"
            ),
            now: Date(timeIntervalSince1970: 1_700_000_000)
        )

        // Proposal's adjective sensitivity bits 6–11 must reflect the target's tier.
        #expect(proposal.adjectiveSensitivity == .restricted,
                "proposal must inherit target drawer's .restricted sensitivity in bits 6–11")
    }

    @Test("propose on a normal target leaves proposal sensitivity at .normal")
    func propose_normalTargetLeavesNormalSensitivity() async throws {
        let dir = FileManager.default.temporaryDirectory
            .appendingPathComponent("locuskit-prop-norm-\(UUID().uuidString)")
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        let path = dir.appendingPathComponent("estate.sqlite3")
        defer { try? FileManager.default.removeItem(at: dir) }

        let estate = try await Estate.create(
            storage: TestStorage.sqlite(path),
            owner: OwnerCredentials(ownerIdentifier: "test-owner")
        )
        let frame = CaptureFrame(
            content: "normal knowledge",
            channel: .typed,
            room: "test-room",
            latticeAnchor: LatticeAnchor(udcCode: "005"),
            addedBy: "test-agent",
            embeddingModelID: "minilm-v1"
            // sensitivity defaults to .normal
        )
        let target = try await estate.capture(frame)
        let proposal = try await estate.propose(
            ProposeFrame(target: target.id, kind: .mutateDrawer, justification: "test"),
            now: Date(timeIntervalSince1970: 1_700_000_000)
        )
        #expect(proposal.adjectiveSensitivity == .normal,
                "proposal against a .normal target must carry .normal sensitivity")
    }

    @Test("propose on a secret target stamps proposal with .secret sensitivity")
    func propose_secretTargetStampsSecret() async throws {
        let dir = FileManager.default.temporaryDirectory
            .appendingPathComponent("locuskit-prop-secret-\(UUID().uuidString)")
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        let path = dir.appendingPathComponent("estate.sqlite3")
        defer { try? FileManager.default.removeItem(at: dir) }

        let estate = try await Estate.create(
            storage: TestStorage.sqlite(path),
            owner: OwnerCredentials(ownerIdentifier: "test-owner")
        )
        let frame = CaptureFrame(
            content: "secret knowledge",
            channel: .typed,
            room: "classified",
            latticeAnchor: LatticeAnchor(udcCode: "005"),
            addedBy: "test-agent",
            embeddingModelID: "minilm-v1",
            sensitivity: .secret
        )
        let target = try await estate.capture(frame)
        let proposal = try await estate.propose(
            ProposeFrame(target: target.id, kind: .mutateDrawer, justification: "test"),
            now: Date(timeIntervalSince1970: 1_700_000_000)
        )
        #expect(proposal.adjectiveSensitivity == .secret,
                "proposal against a .secret target must carry .secret sensitivity")
    }
}

// MARK: - DrawerStore tunnel helper

private extension DrawerStore {

    /// Fetch all non-tombstoned tunnels touching `drawerID` (as source or target).
    /// Test-only helper layered on the public `allTunnels()` surface.
    func tunnelsTouching(_ drawerID: String) async throws -> [Tunnel] {
        let all = try await allTunnels()
        return all.filter { $0.sourceDrawerId == drawerID || $0.targetDrawerId == drawerID }
    }
}

File: packages/kits/LocusKit/rust/tests/sensitivity_inheritance_tests.rs
// sensitivity_inheritance_tests.rs
//
// LocusKit store-layer acceptance tests for sensitivity inheritance (§D.1–D.2).
// Rust twin of Swift `SensitivityInheritanceTests.swift`.
//
// These tests drive the two store-layer primitives directly
// (`consolidate_transactionally`, `fold_in_transactionally`) and the Estate
// verb surface (`propose`) to verify that each produces correctly-stamped
// adjective_bitmap values in bits 6–11 per cookbook §2.3.
//
// ## What is tested
//
// - `consolidate_transactionally`: each `_consolidated_from` tunnel is stamped
//   with max(vague, constituent) adjective sensitivity. Mirrors Swift
//   `testConsolidateTunnelCarriesMaxSensitivity`.
// - `fold_in_transactionally`: the supersedes tunnel carries v2's sensitivity;
//   each `_consolidated_from` tunnel carries max(v2, constituent). Mirrors
//   Swift `testFoldInSupersedesTunnelCarriesV2Sensitivity` and
//   `testFoldInConsolidatedFromTunnelsCarryMaxSensitivity`.
// - `Estate::propose`: the returned proposal's adjective sensitivity (bits
//   6–11) mirrors the target drawer's sensitivity tier. Mirrors Swift
//   `testProposalInheritsTargetSensitivity`.
//
// ## Scope
//
// These tests exercise the LocusKit store and Estate verb layers only — they
// do NOT invoke the GeniusLocusKit coordinator or the consolidation cycle.
// The GLK-layer acceptance tests (including the vague-recall ceiling and
// full-cycle max-sensitivity correctness) live in
// `GeniusLocusKit/rust/tests/sensitivity_inheritance_consolidation_tests.rs`.

use locus_kit::adjectives::AdjectiveSensitivity;
use locus_kit::drawer::Drawer;
use locus_kit::drawer_operational::DrawerFeatureFlags;
use locus_kit::drawer_store::DrawerStore;
use locus_kit::drawer_store_inmemory::InMemoryDrawerStore;
use locus_kit::estate::Estate;
use locus_kit::estate_types::{LatticeAnchor, OwnerCredentials};
use locus_kit::frames::CaptureFrame;
use locus_kit::frames::ProposeFrame;
use locus_kit::drawer_operational::CaptureChannel;
use locus_kit::proposal_operational::ProposalKind;
use std::sync::Arc;
use substrate_kernel::bit_field;
use uuid::Uuid;

const NOW: i64 = 1_700_000_000;
const TEST_PARENT: &str = "00000000-0000-4000-8000-000000000001";

// ─── helpers ──────────────────────────────────────────────────────────────────

fn new_store() -> InMemoryDrawerStore {
    InMemoryDrawerStore::new(NOW, None).expect("store init")
}

fn make_id() -> String {
    Uuid::new_v4().to_string()
}

/// Create a plain constituent drawer with the given sensitivity in bits 6–11
/// of `adjective_bitmap`. Callers must call `store.add_drawer` before using
/// the ID as a constituent in consolidation.
fn constituent_drawer(sensitivity: AdjectiveSensitivity) -> Drawer {
    let mut d = Drawer::new(
        make_id(),
        "constituent content",
        TEST_PARENT,
        "newton",
        NOW,
        "test-model-v1",
    );
    d.udc_code = "001".to_string();
    // Stamp sensitivity into bits 6–11 of adjective_bitmap (cookbook §2.3).
    d.adjective_bitmap = bit_field::write_field(sensitivity.raw_value(), 0i64, 6, 6);
    d
}

/// Create a vague drawer (IS_VAGUE bit set) with the given sensitivity.
fn vague_drawer(sensitivity: AdjectiveSensitivity) -> Drawer {
    let mut d = Drawer::new(
        make_id(),
        "vague cluster",
        TEST_PARENT,
        "newton",
        NOW,
        "test-model-v1",
    );
    d.udc_code = "001".to_string();
    d.operational_bitmap |= DrawerFeatureFlags::IS_VAGUE;
    // Stamp sensitivity into bits 6–11 of adjective_bitmap (cookbook §2.3).
    d.adjective_bitmap = bit_field::write_field(sensitivity.raw_value(), 0i64, 6, 6);
    d
}

/// Return the `_consolidated_from` tunnel ID between a vague drawer and a
/// constituent: `_consolidated_from:{vague_id}:{constituent_id}`.
fn cf_tunnel_id(vague_id: &str, constituent_id: &str) -> String {
    format!("_consolidated_from:{}:{}", vague_id, constituent_id)
}

/// Return the `supersedes` tunnel ID: `supersedes:{v2_id}:{prior_vague_id}`.
fn supersedes_tunnel_id(v2_id: &str, prior_vague_id: &str) -> String {
    format!("supersedes:{}:{}", v2_id, prior_vague_id)
}

// ─── Test 1: consolidate_transactionally — tunnel stamps ─────────────────────

/// `_consolidated_from` tunnels carry max(vague, constituent) sensitivity
/// (§D.1). Each tunnel is checked individually:
/// - vague=Elevated, constituent=Normal  → max = Elevated
/// - vague=Elevated, constituent=Elevated → max = Elevated
/// - vague=Elevated, constituent=Secret  → max = Secret
///
/// Mirrors Swift `testConsolidateTunnelCarriesMaxSensitivity`.
#[test]
fn consolidate_transactionally_tunnel_carries_max_sensitivity() {
    let store = new_store();

    // Three constituents: below-vague, equal, and above-vague sensitivity.
    let c_normal = constituent_drawer(AdjectiveSensitivity::Normal);
    let c_elevated = constituent_drawer(AdjectiveSensitivity::Elevated);
    let c_secret = constituent_drawer(AdjectiveSensitivity::Secret);

    store.add_drawer(&c_normal, NOW).expect("add c_normal");
    store.add_drawer(&c_elevated, NOW).expect("add c_elevated");
    store.add_drawer(&c_secret, NOW).expect("add c_secret");

    // Vague drawer: Elevated sensitivity (floor for all tunnels that touch a
    // normal constituent; ceiling for constituents below Elevated).
    let vague = vague_drawer(AdjectiveSensitivity::Elevated);
    let vague_id = vague.id.clone();
    let constituent_ids: Vec<&str> = vec![&c_normal.id, &c_elevated.id, &c_secret.id];

    store
        .consolidate_transactionally(&vague, &constituent_ids, "newton", NOW)
        .expect("consolidate_transactionally");

    let all_tunnels = store.all_tunnels().expect("all_tunnels");

    // Verify the three _consolidated_from tunnels.
    let check = |constituent_id: &str, expected: AdjectiveSensitivity| {
        let tid = cf_tunnel_id(&vague_id, constituent_id);
        let t = all_tunnels
            .iter()
            .find(|t| t.id == tid)
            .unwrap_or_else(|| panic!("_consolidated_from tunnel {} not found", tid));
        assert_eq!(
            t.adjective_sensitivity(),
            expected,
            "_consolidated_from:{}:{} expected {:?}, got {:?}",
            vague_id,
            constituent_id,
            expected,
            t.adjective_sensitivity()
        );
    };

    // max(Elevated=16, Normal=0)   = Elevated
    check(&c_normal.id, AdjectiveSensitivity::Elevated);
    // max(Elevated=16, Elevated=16) = Elevated
    check(&c_elevated.id, AdjectiveSensitivity::Elevated);
    // max(Elevated=16, Secret=48)  = Secret
    check(&c_secret.id, AdjectiveSensitivity::Secret);
}

/// When vague sensitivity exceeds every constituent, all tunnels carry the
/// vague tier. Mirrors Swift `testConsolidateTunnelFlooredByVagueSensitivity`.
#[test]
fn consolidate_transactionally_vague_is_floor_when_higher() {
    let store = new_store();

    // All constituents: Normal.
    let c1 = constituent_drawer(AdjectiveSensitivity::Normal);
    let c2 = constituent_drawer(AdjectiveSensitivity::Normal);
    let c3 = constituent_drawer(AdjectiveSensitivity::Normal);

    store.add_drawer(&c1, NOW).expect("add c1");
    store.add_drawer(&c2, NOW).expect("add c2");
    store.add_drawer(&c3, NOW).expect("add c3");

    // Vague: Secret — all tunnels must be floored at Secret.
    let vague = vague_drawer(AdjectiveSensitivity::Secret);
    let vague_id = vague.id.clone();
    let constituent_ids = vec![c1.id.as_str(), c2.id.as_str(), c3.id.as_str()];

    store
        .consolidate_transactionally(&vague, &constituent_ids, "newton", NOW)
        .expect("consolidate_transactionally");

    let all_tunnels = store.all_tunnels().expect("all_tunnels");

    for cid in &[&c1.id, &c2.id, &c3.id] {
        let tid = cf_tunnel_id(&vague_id, cid);
        let t = all_tunnels
            .iter()
            .find(|t| t.id == tid)
            .unwrap_or_else(|| panic!("tunnel {} not found", tid));
        assert_eq!(
            t.adjective_sensitivity(),
            AdjectiveSensitivity::Secret,
            "tunnel {} should be Secret (vague floor), got {:?}",
            tid,
            t.adjective_sensitivity()
        );
    }
}

// ─── Test 2: fold_in_transactionally — tunnel stamps ─────────────────────────

/// `fold_in_transactionally` stamps:
/// - the `supersedes` tunnel with v2's sensitivity
/// - each `_consolidated_from` tunnel with max(v2, constituent) sensitivity
///
/// Mirrors Swift `testFoldInSupersedesTunnelCarriesV2Sensitivity` and
/// `testFoldInConsolidatedFromTunnelsCarryMaxSensitivity`.
#[test]
fn fold_in_transactionally_tunnel_stamps() {
    let store = new_store();

    // Three constituents: Normal, Elevated, Secret.
    let c_normal = constituent_drawer(AdjectiveSensitivity::Normal);
    let c_elevated = constituent_drawer(AdjectiveSensitivity::Elevated);
    let c_secret = constituent_drawer(AdjectiveSensitivity::Secret);

    store.add_drawer(&c_normal, NOW).expect("add c_normal");
    store.add_drawer(&c_elevated, NOW).expect("add c_elevated");
    store.add_drawer(&c_secret, NOW).expect("add c_secret");

    // Prior vague: Normal sensitivity. Add directly to store.
    let prior_vague = vague_drawer(AdjectiveSensitivity::Normal);
    store.add_drawer(&prior_vague, NOW).expect("add prior_vague");
    let prior_vague_id = prior_vague.id.clone();
    let shared_lineage = prior_vague.lineage_id; // v2 must share this

    // v2: Restricted sensitivity; shares lineage with prior vague.
    // fold_in_transactionally checks that vague_v2.lineage_id == prior.lineageID.
    let mut vague_v2 = vague_drawer(AdjectiveSensitivity::Restricted);
    vague_v2.lineage_id = shared_lineage;
    let v2_id = vague_v2.id.clone();

    let enlarged_ids = vec![
        c_normal.id.as_str(),
        c_elevated.id.as_str(),
        c_secret.id.as_str(),
    ];

    store
        .fold_in_transactionally(&vague_v2, &prior_vague_id, &enlarged_ids, "newton", NOW)
        .expect("fold_in_transactionally");

    let all_tunnels = store.all_tunnels().expect("all_tunnels");

    // supersedes tunnel: v2.sensitivity = Restricted.
    let sup_id = supersedes_tunnel_id(&v2_id, &prior_vague_id);
    let sup = all_tunnels
        .iter()
        .find(|t| t.id == sup_id)
        .unwrap_or_else(|| panic!("supersedes tunnel {} not found", sup_id));
    assert_eq!(
        sup.adjective_sensitivity(),
        AdjectiveSensitivity::Restricted,
        "supersedes tunnel must carry v2 sensitivity (Restricted), got {:?}",
        sup.adjective_sensitivity()
    );

    // _consolidated_from tunnels: max(v2=Restricted, constituent).
    let check_cf = |constituent_id: &str, expected: AdjectiveSensitivity| {
        let tid = cf_tunnel_id(&v2_id, constituent_id);
        let t = all_tunnels
            .iter()
            .find(|t| t.id == tid)
            .unwrap_or_else(|| panic!("_consolidated_from tunnel {} not found", tid));
        assert_eq!(
            t.adjective_sensitivity(),
            expected,
            "_consolidated_from:{}:{} expected {:?}, got {:?}",
            v2_id,
            constituent_id,
            expected,
            t.adjective_sensitivity()
        );
    };

    // max(Restricted=32, Normal=0)    = Restricted
    check_cf(&c_normal.id, AdjectiveSensitivity::Restricted);
    // max(Restricted=32, Elevated=16) = Restricted
    check_cf(&c_elevated.id, AdjectiveSensitivity::Restricted);
    // max(Restricted=32, Secret=48)   = Secret
    check_cf(&c_secret.id, AdjectiveSensitivity::Secret);
}

// ─── Test 3: Estate::propose stamps proposal adjective sensitivity ────────────

/// A proposal inherits its target drawer's sensitivity tier in bits 6–11 of
/// `adjective_bitmap` (§D.2). Three tiers verified: Normal, Elevated, Secret.
///
/// Mirrors Swift `testProposalInheritsTargetSensitivity`.
#[test]
fn propose_inherits_target_drawer_sensitivity() {
    // Use all three distinct tiers to verify the proposal stamps each one.
    for sensitivity in [
        AdjectiveSensitivity::Normal,
        AdjectiveSensitivity::Elevated,
        AdjectiveSensitivity::Secret,
    ] {
        let store = Arc::new(new_store()) as Arc<dyn DrawerStore>;
        let estate = Estate::create(store, OwnerCredentials::new("newton"), None)
            .expect("create estate");

        // Capture a drawer with the given sensitivity tier.
        let mut frame = CaptureFrame::new(
            "target content for propose",
            CaptureChannel::Typed,
            "inbox",
            LatticeAnchor::udc("001"),
            "newton",
            "test-model-v1",
        );
        frame.sensitivity = sensitivity;
        let drawer = estate.capture(frame, NOW).expect("capture");

        // Propose against that drawer.
        let proposal = estate
            .propose(ProposeFrame::new(&drawer.id, ProposalKind::MutateDrawer), NOW)
            .expect("propose");

        assert_eq!(
            proposal.adjective_sensitivity(),
            sensitivity,
            "proposal must inherit target sensitivity {:?} (bits 6–11), got {:?}",
            sensitivity,
            proposal.adjective_sensitivity()
        );
    }
}

/// A proposal filed against a Normal (raw 0) target stays Normal even though
/// zero is the default for uninitialised bitmaps — ensuring the stamp path is
/// not accidentally elided for the zero case.
///
/// Mirrors Swift `testProposalNormalTargetStaysNormal`.
#[test]
fn propose_normal_target_stays_normal() {
    let store = Arc::new(new_store()) as Arc<dyn DrawerStore>;
    let estate = Estate::create(store, OwnerCredentials::new("newton"), None)
        .expect("create estate");

    // Capture a Normal-sensitivity drawer (the default).
    let frame = CaptureFrame::new(
        "normal sensitivity target",
        CaptureChannel::Typed,
        "inbox",
        LatticeAnchor::udc("001"),
        "newton",
        "test-model-v1",
    );
    let drawer = estate.capture(frame, NOW).expect("capture");

    let proposal = estate
        .propose(ProposeFrame::new(&drawer.id, ProposalKind::MutateDrawer), NOW)
        .expect("propose");

    assert_eq!(
        proposal.adjective_sensitivity(),
        AdjectiveSensitivity::Normal,
        "proposal against Normal target must stay Normal, got {:?}",
        proposal.adjective_sensitivity()
    );
}
